README.md
# solr-node
Simple Solr Node Client Project

[![NPM](https://nodei.co/npm/solr-node.png?downloads=true&stars=true)](https://nodei.co/npm/solr-node/)

## Install

```
npm install solr-node
```


## Usage
- Client: http://www.godong9.com/solrnode/Client.html
- Query: http://www.godong9.com/solrnode/Query.html

### Create Client
```js
// Require module
var SolrNode = require('solr-node');

// Create client
var client = new SolrNode({
    host: '127.0.0.1',
    port: '8983',
    core: 'test',
    protocol: 'http'
});

// Set Debug Level
var client = new SolrNode({
    host: '127.0.0.1',
    port: '8983',
    core: 'test',
    protocol: 'http',
    debugLevel: 'ERROR' // log4js debug level paramter
});
```

### Search

```js
// Create query
var strQuery = client.query().q('text:test');
var objQuery = client.query().q({text:'test', title:'test'});
var myStrQuery = 'q=text:test&wt=json';

// Search documents using strQuery
solrClient.search(strQuery, function (err, result) {
   if (err) {
      console.log(err);
      return;
   }
   console.log('Response:', result.response);
});

// Search documents using objQuery
solrClient.search(objQuery, function (err, result) {
   if (err) {
      console.log(err);
      return;
   }
   console.log('Response:', result.response);
});

// Search documents using myStrQuery
solrClient.search(myStrQuery, function (err, result) {
   if (err) {
      console.log(err);
      return;
   }
   console.log('Response:', result.response);
});

```

### Update

```js
// JSON Data
var data = {
    text: 'test',
    title: 'test'
};

// Update document to Solr server
client.update(data, function(err, result) {
   if (err) {
      console.log(err);
      return;
   }
   console.log('Response:', result.responseHeader);
});

```

### Delete

```js
// Delete Query
var strQuery = 'id:testid'
var objQuery = {id:'testid'}

// Delete document using strQuery
client.delete(strQuery, function(err, result) {
   if (err) {
      console.log(err);
      return;
   }
   console.log('Response:', result.responseHeader);
});

// Delete document using objQuery
client.delete(objQuery, function(err, result) {
   if (err) {
      console.log(err);
      return;
   }
   console.log('Response:', result.responseHeader);
});

```

## Test & Coverage & Docs

```js
gulp 
```


HISTORY.md
### 0.0.5
    * Support search, terms, update, delete
     
### 0.1.0
    * Support spell
    * Support commit
    * Support ping
    * Support mlt
    
### 0.1.1
    * Support facet
    * Fix #23 (When query is object, make invalid url.)

### 0.1.2
    * Fix #25 (requestPost url bug)
    * Support group
    * Support hl
    
### 0.1.3
    * Fix #30 ("&wt=json" don't need when query is String.)
   
### 1.0.0
    * Fix #33 (Fix comments)
    * Update #34 (Add error exception)
    * Support #35 (wt)
    * Update #38 (Add test case)
    * Update #40 (Update client.js) 
     
### 1.0.1
    * Add docs
    * Add docs generator (gulp doc)
   
### 1.0.2
    * Support set debug level (#45)

### 1.0.3
    * Support #47
    
### 1.0.4
    * Fix #49
    
### 1.0.5
    * Support #50
    
### 1.0.6
    * Support #51
    
### 1.0.7
    * Support #53
    
### 1.0.8
    * Support #55
    
### 1.0.9
    * Fix #56
    
### 1.0.10
    * Fix #58
    
### 1.0.11
    * Fix client
    
### 1.0.12
    * Fix client

test/client-test.js
/**
 * Created by godong on 2016. 3. 9..
 */

/**
 * Require modules
 */
var expect = require('chai').expect,
  nock = require('nock'),
  Client = require('../lib/client');

describe('Client', function() {
  var testClient = new Client({core: 'test'});
  before(function() {
    nock.disableNetConnect();
    nock.enableNetConnect(/127.0.0.1/);
    nock.cleanAll();
    nock('http://127.0.0.1:8983')
      .persist()
      .filteringPath(/\/solr\/error\/select.*/g, '/error/select/mock')
      .get('/error/select/mock')
      .reply(404);

    nock('http://127.0.0.1:8983')
      .persist()
      .filteringPath(/\/solr\/error\/update.*/g, '/error/update/mock')
      .post('/error/update/mock')
      .reply(404);

    nock('http://127.0.0.1:8983')
      .persist()
      .get('/solr/test/select?q=text:test&facet=true&facet.query=test&wt=json')
      .reply(200, JSON.stringify({
        responseHeader: {
          status: 0,
          QTime: 86,
          params: { q: 'text:test', wt: 'json' }
        },
        response:{
          numFound: 10,
          start: 0,
          docs: []
        },
        facet_counts:{
          facet_queries: {
            test: 18
          }
        }
      }));

    nock('http://127.0.0.1:8983')
      .persist()
      .filteringPath(/\/solr\/test\/select.*/g, '/test/select/mock')
      .get('/test/select/mock')
      .reply(200, JSON.stringify({
        responseHeader: {
          status: 0,
          QTime: 86,
          params: { q: 'text:test', wt: 'json' }
        },
        response:{
          numFound: 10,
          start: 0,
          docs: []
        }
      }));

    nock('http://127.0.0.1:8983')
      .persist()
      .filteringPath(/\/solr\/test\/terms.*/g, '/test/terms/mock')
      .get('/test/terms/mock')
      .reply(200, JSON.stringify({
        responseHeader: {
          status: 0,
          QTime: 86
        },
        terms:{
          text: []
        }
      }));

    nock('http://127.0.0.1:8983')
      .persist()
      .filteringPath(/\/solr\/test\/mlt.*/g, '/test/mlt/mock')
      .get('/test/mlt/mock')
      .reply(200, JSON.stringify({
        responseHeader: {
          status: 0,
          QTime: 86
        },
        response: {
          numFound: 10,
          start: 0,
          docs: []
        },
        moreLikeThis: {
          "testId1": {
            numFound: 10,
            start: 0,
            docs: []
          },
          "testId2": {
            numFound: 10,
            start: 0,
            docs: []
          }
        }
      }));

    nock('http://127.0.0.1:8983')
      .persist()
      .filteringPath(/\/solr\/test\/spell.*/g, '/test/spell/mock')
      .get('/test/spell/mock')
      .reply(200, JSON.stringify({
        responseHeader: {
          status: 0,
          QTime: 86
        },
        response: {
          numFound: 10,
          start: 0,
          docs: []
        },
        spellcheck:{
          suggestions: [],
          collations: []
        }
      }));

    nock('http://127.0.0.1:8983')
      .persist()
      .filteringPath(/\/solr\/test\/update.*/g, '/test/update/mock')
      .post('/test/update/mock')
      .reply(200, JSON.stringify({
        responseHeader: {
          status: 0,
          QTime: 86
        }
      }));

    nock('http://127.0.0.1:8983')
      .persist()
      .filteringPath(/\/solr\/test\/admin\/ping.*/g, '/test/admin/ping/mock')
      .get('/test/admin/ping/mock')
      .reply(200, JSON.stringify({
        responseHeader: {
          status: 0,
          QTime: 2
        }
      }));

  });

  after(function() {
    nock.cleanAll();
  });

  describe('#constructor', function() {
    it('should create default client.', function() {
      //given
      var options = {};
      //when
      var client = new Client(options);
      //then
      expect(client.options).to.eql({
        host: '127.0.0.1',
        port: '8983',
        core: '',
        rootPath: 'solr',
        protocol: 'http'
      });
    });

    it('should create client when core:"test".', function() {
      //given
      var options = {
        core: 'test',
        debugLevel: 'ALL'
      };
      //when
      var client = new Client(options);
      //then
      expect(client.options).to.eql({
        host: '127.0.0.1',
        port: '8983',
        core: 'test',
        rootPath: 'solr',
        protocol: 'http'
      });
    });
  });

  describe('#_makeHostUrl', function() {
    it('should get host url.', function() {
      //given
      var protocol = 'http';
      var host = '127.0.0.1';
      var port = '8983';
      //when
      var hostUrl = testClient._makeHostUrl(protocol, host, port);
      //then
      expect(hostUrl).to.equal('http://127.0.0.1:8983');
    });

    it('should get host url when port is empty and protocol is https.', function() {
      //given
      var protocol = 'https';
      var host = 'test.com';
      var port = '';
      //when
      var hostUrl = testClient._makeHostUrl(protocol, host, port);
      //then
      expect(hostUrl).to.equal('https://test.com');
    });
  });

  describe('#_requestGet', function() {
    it('should get search error from server.', function(done) {
      //given
      var client = new Client({core: 'error'});
      //when
      client._requestGet(client.SEARCH_PATH, "q=*:*", function(err, result) {
        //then
        expect(err).to.equal('Solr server error: 404');
        expect(result).to.not.exist;
        done();
      });
    });

    it('should get notes data from server when query string.', function(done) {
      //given
      var client = new Client({core: 'test'});
      //when
      client._requestGet(client.SEARCH_PATH, 'q=text:text&wt=json', function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.response).to.exist;
        done();
      });
    });

    it('should get notes data from server when query is query instance.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query = client.query().q('text:test');
      //when
      client._requestGet(client.SEARCH_PATH, query, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.response).to.exist;
        done();
      });
    });

    it('should get notes data from server when query is query instance but query is not string.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query = client.query().q(null);
      //when
      client._requestGet(client.SEARCH_PATH, query, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.response).to.exist;
        done();
      });
    });

    it('should get notes data from server when query is string.', function(done) {
      //given
      var client = new Client({core: 'test'});
      //when
      client._requestGet(client.SEARCH_PATH, "q=*:*", function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.response).to.exist;
        done();
      });
    });

    it('should get notes data from server when query is null.', function(done) {
      //given
      var client = new Client({core: 'test'});
      //when
      client._requestGet(client.SEARCH_PATH, null, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.response).to.exist;
        done();
      });
    });
  });

  describe('#_requestPost', function() {
    it('should get post error from server.', function(done) {
      //given
      var client = new Client({core: 'error'});
      //when
      client._requestPost(client.UPDATE_PATH, {}, {}, function(err, result) {
        //then
        expect(err).to.equal('Solr server error: 404');
        expect(result).to.not.exist;
        done();
      });
    });

    it('should post data to server when options is null.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var data = {
        test: 'test',
        title: 'test'
      };
      var options = null;
      //when
      client._requestPost(client.UPDATE_PATH, data, options, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.responseHeader).to.exist;
        done();
      });
    });

    it('should post data to server when options is object.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var data = {
        test: 'test',
        title: 'test'
      };
      //when
      client._requestPost(client.UPDATE_PATH, data, {}, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.responseHeader).to.exist;
        done();
      });
    });

    it('should post data to server when options is string.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var data = {
        test: 'test',
        title: 'test'
      };
      //when
      client._requestPost(client.UPDATE_PATH, data, 'commit=true', function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.responseHeader).to.exist;
        done();
      });
    });
  });

  describe('#search', function() {
    it('should search data when not using query.', function(done) {
      //given
      var client = new Client({core: 'test'});
      //when
      client.search('q=text:test', function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.response).to.exist;
        done();
      });
    });

    it('should search data when query is string.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query = client.query().q('text:test');
      //when
      client.search(query, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.response).to.exist;
        done();
      });
    });

    it('should search data when query is object.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query =
        client.query()
          .q({text:'test', title:'test'});
      //when
      client.search(query, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.response).to.exist;
        done();
      });
    });

    it('should search data with query options.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query =
        client.query()
          .q({text:'test', title:'test'})
          .fl(['text', 'title'])
          .start(0)
          .rows(10)
          .sort({like:'desc'})
          .fq({field:'hate', value:0});

      //when
      client.search(query, function(err, result) {
        //then
        expect(query.params).to.eql([
          'q=text:test%20AND%20title:test',
          'fl=text,title',
          'start=0',
          'rows=10',
          'sort=like desc',
          'fq=hate:0'
        ]);

        expect(err).to.not.exist;
        expect(result.response).to.exist;
        done();
      });
    });

    it('should search data when query and facet query are exist.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query =
        client.query()
          .q({text:'test'})
          .facetQuery({query:'test'});

      //when
      client.search(query, function(err, result) {
        //then
        expect(query.params).to.eql([
          'q=text:test',
          'facet=true',
          'facet.query=test'
        ]);

        expect(err).to.not.exist;
        expect(result.response).to.exist;
        expect(result.facet_counts).to.exist;
        done();
      });
    });
  });

  describe('#terms', function() {
    it('should get terms data.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query =
        client.query()
          .termsQuery({
            fl: 'text',
            prefix: 'te'
          });
      //when
      client.terms(query, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.terms).to.exist;
        done();
      });
    });
  });

  describe('#mlt', function() {
    it('should get mlt data.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query =
        client.query()
          .q({title: 'test'})
          .mltQuery({
            fl: ['title', 'text'],
            mindf: 1,
            mintf: 1
          });
      //when
      client.mlt(query, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.response).to.exist;
        expect(result.moreLikeThis).to.exist;
        done();
      });
    });
  });

  describe('#spell', function() {
    it('should get spell data.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query =
        client.query()
          .df('title')
          .spellcheckQuery({
            q: 'tes',
            build: true,
            collate: true,
            maxCollations: 5,
            maxCollationTries: 3,
            maxCollationEvaluations: 2,
            collateExtendedResults: true,
            collateMaxCollectDocs: 5,
            dictionary: "defaule",
            extendedResults: true,
            onlyMorePopular: true,
            maxResultsForSuggest: 5,
            alternativeTermCount: 3,
            reload: true,
            accuracy: 1.0
          });
      //when
      client.spell(query, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.spellcheck).to.exist;
        done();
      });
    });
  });

  describe('#update', function() {
    it('should post data to server without options.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var data = {
        text: 'test',
        title: 'test'
      };
      //when
      client.update(data, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.responseHeader).to.exist;
        done();
      });
    });

    it('should post data to server with options.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var data = {
        text: 'test',
        title: 'test'
      };
      var options = { commit: true };
      //when
      client.update(data, options, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.responseHeader).to.exist;
        done();
      });
    });
  });

  describe('#delete', function() {
    it('should delete data to server without options.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query = {id:'testid1'};
      //when
      client.delete(query, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.responseHeader).to.exist;
        done();
      });
    });

    it('should delete data to server with options when query is object.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query = {id:'testid2'};
      var options = { commit: true };
      //when
      client.delete(query, options, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.responseHeader).to.exist;
        done();
      });
    });

    it('should delete data to server with options when query is string.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query = "id:testid3";
      var options = { commit: true };
      //when
      client.delete(query, options, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.responseHeader).to.exist;
        done();
      });
    });

    it('should delete data to server with options when query is null.', function(done) {
      //given
      var client = new Client({core: 'test'});
      var query = null;
      var options = { commit: true };
      //when
      client.delete(query, options, function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.responseHeader).to.exist;
        done();
      });
    });
  });

  describe('#ping', function() {
    it('should request ping.', function(done) {
      //given
      var client = new Client({core: 'test'});
      //when
      client.ping(function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.responseHeader).to.exist;
        done();
      });
    });
  });

  describe('#commit', function() {
    it('should request commit.', function(done) {
      //given
      var client = new Client({core: 'test'});
      //when
      client.commit(function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.responseHeader).to.exist;
        done();
      });
    });
  });

  describe('#softCommit', function() {
    it('should request softCommit.', function(done) {
      //given
      var client = new Client({core: 'test'});
      //when
      client.softCommit(function(err, result) {
        //then
        expect(err).to.not.exist;
        expect(result.responseHeader).to.exist;
        done();
      });
    });
  });

});



gulpfile.js
var gulp = require('gulp'),
  istanbul = require('gulp-istanbul'),
  jshint = require('gulp-jshint'),
  stylish = require('jshint-stylish'),
  mocha = require('gulp-mocha');

gulp.task('lint', function() {
  return gulp.src('./lib/**/*.js')
    .pipe(jshint({}))
    .pipe(jshint.reporter(stylish));
});

gulp.task('pre-test', function () {
  return gulp.src(['./lib/**/*.js'])
    // Covering files
    .pipe(istanbul())
    // Force `require` to return covered files
    .pipe(istanbul.hookRequire());
});

gulp.task('test', ['pre-test'], function() {
  return gulp.src('./test/**/*.js', {read: false})
    .pipe(mocha({reporter: 'nyan'}))
    // Creating the reports after tests ran
    .pipe(istanbul.writeReports())
    // Enforce a coverage of at least 90%
    .pipe(istanbul.enforceThresholds({ thresholds: { global: 90 } }));
});

gulp.task('default', ['lint', 'test']);

lib/client.js
/**
 * Created by godong on 2016. 3. 9..
 */

/**
 * Require modules
 */
var _ = require('underscore'),
  request = require('request'),
  logger =  require('log4js').getLogger('sole-node'),
  QueryString = require('querystring'),
  Query = require('./query');

/**
 * Solr Node Client
 *
 * @constructor
 *
 * @param {Object} options
 * @param {String} [options.host] - host address of Solr server
 * @param {Number|String} [options.port] - port number of Solr server
 * @param {String} [options.core] - client core name
 * @param {String} [options.rootPath] - solr root path
 * @param {String} [options.protocol] - request protocol ('http'|'https')
 * @param {String} [options.debugLevel] - log4js debug level ('ALL'|'DEBUG'|'INFO'|'ERROR'|...)
 */
function Client(options) {
  this.options = {
    host: options.host || '127.0.0.1',
    port: options.port || '8983',
    core: options.core || '',
    rootPath: options.rootPath || 'solr',
    protocol: options.protocol || 'http'
  };

  if (options.debugLevel) {
    logger.setLevel(options.debugLevel);
  }

  // Path Constants List
  this.SEARCH_PATH = 'select';
  this.TERMS_PATH = 'terms';
  this.SPELL_PATH = 'spell';
  this.MLT_PATH = 'mlt';
  this.UPDATE_PATH = 'update';
  this.PING_PATH = 'admin/ping';
}

/**
 * Make host url
 * @private
 *
 * @param {String} protocol - protocol ('http'|'https')
 * @param {String} host - host address
 * @param {String|Number} port - port number
 *
 * @returns {String}
 */
Client.prototype._makeHostUrl = function(protocol, host, port) {
  if (port) {
    return protocol + '://' + host + ':' + port;
  } else {
    return protocol + '://' + host;
  }
};

/**
 * Request get
 *
 * @param {String} path - target path
 * @param {Object|String} query - query
 * @param {Function} finalCallback - (err, result)
 */
Client.prototype._requestGet = function(path, query, finalCallback) {
  var params, options, requestPrefixUrl, requestFullPath;

  if (query instanceof Query) {
    params = query.toString();
  } else if (_.isString(query)) {
    params = query;
  } else {
    params = 'q=*:*';
  }
  requestPrefixUrl = this._makeHostUrl(this.options.protocol, this.options.host, this.options.port);
  requestPrefixUrl += '/' + [this.options.rootPath, this.options.core, path].join('/');

  requestFullPath = requestPrefixUrl + '?' + params;

  logger.debug('[_requestGet] requestFullPath: ', requestFullPath);

  options = {
    method: 'GET',
    url: requestFullPath,
    headers: {
      'accept' : 'application/json; charset=utf-8'
    }
  };
  request(options, function (err, res, body) {
    var result;
    /* istanbul ignore next */
    if (err) {
      return finalCallback(err);
    }
    if (res.statusCode !== 200) {
      logger.error(body);
      return finalCallback('Solr server error: ' + res.statusCode);
    }
    /* istanbul ignore next */
    try {
      result = JSON.parse(body);
    } catch(e) {
      result = body;
    }
    return finalCallback(null, result);
  });
};

/**
 * Request post
 *
 * @param {String} path - target path
 * @param {Object} data - json data
 * @param {Object|String} urlOptions - url options
 * @param {Function} finalCallback - (err, result)
 */
Client.prototype._requestPost = function(path, data, urlOptions, finalCallback) {
  var params, options, requestPrefixUrl, requestFullPath;

  if (_.isString(urlOptions)) {
    params = urlOptions;
  } else if (_.isObject(urlOptions)) {
    params = QueryString.stringify(urlOptions);
  } else {
    params = '';
  }

  requestPrefixUrl = this._makeHostUrl(this.options.protocol, this.options.host, this.options.port);
  requestPrefixUrl += '/' + [this.options.rootPath, this.options.core, path].join('/');

  requestFullPath = requestPrefixUrl + '?' + params;

  logger.debug('[_requestPost] requestFullPath: ', requestFullPath);
  logger.debug('[_requestPost] data: ', data);

  options = {
    method: 'POST',
    url: requestFullPath,
    json: data,
    headers: {
      'accept' : 'application/json; charset=utf-8',
      'content-type' : 'application/json; charset=utf-8'
    }
  };
  request(options, function (err, res, body) {
    var result;
    /* istanbul ignore next */
    if (err) {
      return finalCallback(err);
    }
    if (res.statusCode !== 200) {
      logger.error(body);
      return finalCallback('Solr server error: ' + res.statusCode);
    }
    /* istanbul ignore next */
    try {
      result = JSON.parse(body);
    } catch(e) {
      result = body;
    }
    return finalCallback(null, result);
  });
};

/**
 * Make Query instance and return
 *
 * @returns {Object}
 */
Client.prototype.query = function() {
  return new Query();
};

/**
 * Search
 *
 * @param {Object|String} query
 * @param {Function} finalCallback - (err, result)
 */
Client.prototype.search = function(query, finalCallback) {
  this._requestGet(this.SEARCH_PATH, query, finalCallback);
};

/**
 * Terms
 *
 * @param {Object|String} query
 * @param {Function} finalCallback - (err, result)
 */
Client.prototype.terms = function(query, finalCallback) {
  this._requestGet(this.TERMS_PATH, query, finalCallback);
};

/**
 * Mlt
 *
 * @param {Object|String} query
 * @param {Function} finalCallback - (err, result)
 */
Client.prototype.mlt = function(query, finalCallback) {
  this._requestGet(this.MLT_PATH, query, finalCallback);
};

/**
 * Spell
 *
 * @param {Object|String} query
 * @param {Function} finalCallback - (err, result)
 */
Client.prototype.spell = function(query, finalCallback) {
  this._requestGet(this.SPELL_PATH, query, finalCallback);
};

/**
 * Update
 *
 * @param {Object} data - json data
 * @param {Object|Function} [options] - update options
 * @param {Function} finalCallback - (err, result)
 */
Client.prototype.update = function(data, options, finalCallback) {
  var bodyData;
  if (arguments.length < 3 && _.isFunction(options)) {
    finalCallback = options;
    options = { commit: true }; // 'commit:true' option is default
  }
  bodyData = {
    add: {
      doc: data,
      overwrite: true
    }
  };
  this._requestPost(this.UPDATE_PATH, bodyData, options, finalCallback);
};

/**
 * Delete
 *
 * @param {String|Object} query - query
 * @param {Object|Function} [options] - delete options
 * @param {Function} finalCallback - (err, result)
 */
Client.prototype.delete = function(query, options, finalCallback) {
  var bodyData, bodyQuery;
  if (arguments.length < 3 && _.isFunction(options)) {
    finalCallback = options;
    options = { commit: true }; // 'commit:true' option is default
  }

  if (_.isString(query)) {
    bodyQuery = query;
  } else if (_.isObject(query)) {
    bodyQuery = QueryString.stringify(query, ' AND ', ':');
  } else {
    bodyQuery = '';
  }

  bodyData = {
    'delete': {
      query: bodyQuery
    }
  };

  this._requestPost(this.UPDATE_PATH, bodyData, options, finalCallback);
};

/**
 * Ping
 *
 * @param {Function} finalCallback - (err, result)
 */
Client.prototype.ping = function(finalCallback) {
  this._requestGet(this.PING_PATH, '', finalCallback);
};

/**
 * Commit
 *
 * @param {Function} finalCallback - (err, result)
 */
Client.prototype.commit = function(finalCallback) {
  this._requestPost(this.UPDATE_PATH, {}, {commit: true}, finalCallback);
};

/**
 * SoftCommit
 *
 * @param {Function} finalCallback - (err, result)
 */
Client.prototype.softCommit = function(finalCallback) {
  this._requestPost(this.UPDATE_PATH, {}, {softCommit: true}, finalCallback);
};


module.exports = Client;